mono/mini/simd-intrinsics-netcore.c

```diff
@@ -0,0 +1,438 @@
+				  mono_metadata_type_equal (fsig->params [1], type));
+		ins = emit_simd_ins (cfg, klass, OP_XEQUAL, args [0]->dreg, args [1]->dreg);
+		if (id == SN_op_Inequality) {
+			int sreg = ins->dreg;
+			int dreg = alloc_ireg (cfg);
+			EMIT_NEW_UNALU (cfg, ins, OP_INOT, dreg, sreg);
+		}
+		return ins;
+	case SN_GreaterThan:
+	case SN_LessThan:
+		g_assert (fsig->param_count == 2 && mono_metadata_type_equal (fsig->ret, type) && mono_metadata_type_equal (fsig->params [0], type) && mono_metadata_type_equal (fsig->params [1], type));
+		is_unsigned = etype->type == MONO_TYPE_U1 || etype->type == MONO_TYPE_U2 || etype->type == MONO_TYPE_U4 || etype->type == MONO_TYPE_U8;
+		ins = emit_xcompare (cfg, klass, etype, args [0], args [1]);
+		switch (id) {
+		case SN_GreaterThan:
+			ins->inst_c0 = is_unsigned ? CMP_GT_UN : CMP_GT;
+			break;
+		case SN_LessThan:
+			ins->inst_c0 = is_unsigned ? CMP_LT_UN : CMP_LT;
+			break;
+		default:
+			g_assert_not_reached ();
+		}
+		return ins;
+	case SN_op_Explicit:
+		return emit_simd_ins (cfg, klass, OP_XMOVE, args [0]->dreg, -1);
+	case SN_op_Addition:
+	case SN_op_Subtraction:
+	case SN_op_Division:
+	case SN_op_Multiply:
+	case SN_op_BitwiseAnd:
+	case SN_op_BitwiseOr:
+	case SN_op_ExclusiveOr:
+		if (!(fsig->param_count == 2 && mono_metadata_type_equal (fsig->ret, type) && mono_metadata_type_equal (fsig->params [0], type) && mono_metadata_type_equal (fsig->params [1], type)))
+			return NULL;
+		ins = emit_simd_ins (cfg, klass, OP_XBINOP, args [0]->dreg, args [1]->dreg);
+		if (etype->type == MONO_TYPE_R4 || etype->type == MONO_TYPE_R8) {
+			switch (id) {
+			case SN_op_Addition:
+				ins->inst_c0 = OP_FADD;
+				break;
+			case SN_op_Subtraction:
+				ins->inst_c0 = OP_FSUB;
+				break;
+			case SN_op_Multiply:
+				ins->inst_c0 = OP_FMUL;
+				break;
+			case SN_op_Division:
+				ins->inst_c0 = OP_FDIV;
+				break;
+			default:
+				NULLIFY_INS (ins);
+				return NULL;
+			}
+		} else {
+			switch (id) {
+			case SN_op_Addition:
+				ins->inst_c0 = OP_IADD;
+				break;
+			case SN_op_Subtraction:
+				ins->inst_c0 = OP_ISUB;
+				break;
+				/*
+			case SN_op_Division:
+				ins->inst_c0 = OP_IDIV;
+				break;
+			case SN_op_Multiply:
+				ins->inst_c0 = OP_IMUL;
+				break;
+				*/
+			case SN_op_BitwiseAnd:
+				ins->inst_c0 = OP_IAND;
+				break;
+			case SN_op_BitwiseOr:
+				ins->inst_c0 = OP_IOR;
+				break;
+			case SN_op_ExclusiveOr:
+				ins->inst_c0 = OP_IXOR;
+				break;
+			default:
+				NULLIFY_INS (ins);
+				return NULL;
+			}
+		}
+		return ins;
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+MonoInst*
+mono_emit_simd_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fsig, MonoInst **args)
+{
+	const char *class_name;
+	const char *class_ns;
+	MonoImage *image = m_class_get_image (cmethod->klass);
+
+	if (image != mono_get_corlib ())
+		return NULL;
+	if (!COMPILE_LLVM (cfg))
+		return NULL;
+	// FIXME:
+	if (cfg->compile_aot)
+		return NULL;
```

Comment:
Why is it disabled for AOT btw? I guess it's ok to do it for now but a comment with explanation would be very welcome.

Reply:
Depends on the current CPU is why?

Reply:
Isn't SSE2 minimal requirement for x86_64?

Reply:
I'd expect the LLVM CPU selection should kick in but maybe it doesn't fallback quite well with current settings. Anyway, just guessing.

This makes it quite hard to test the impact. Running performance tests with `--llvm` is non-starter at the moment, it is too slow and has low test coverage. I do run performance pipeline with `-mcpu=native` for LLVM AOT and would like to see the impact there.

Reply:
Automatic selection is troublesome for AOT and copy binaries to other system.
Yes SSE2 is required for amd64.

Reply:
@jaykrell I understand that. That's why there's some `-mcpu=generic` default which I though would still work for this. It can be overridden if desired (eg. doing AOT on deployment).